extensions/default/src/MergeDataSource/index.js

```diff
@@ -0,0 +1,140 @@
+import { DicomMetadataStore, IWebApiDataSource } from '@ohif/core';
+import { executeFunction, getKeyByLevel } from './utils';
+
+function createMergeDataSourceApi(mergeConfig, UserAuthenticationService, extensionManager) {
+  const { seriesMerge } = mergeConfig;
+  const { dataSourceNames } = seriesMerge;
+
+  const callForAllDataSourcesAsync = async (path, args) => {
+    const mergeMap = {
+      'query.studies.search': {
+        mergeKey: 'studyInstanceUid',
+        level: 0,
+        /** TODO */
+        tagFunc: x => x,
+      },
+      'retrieve.series.metadata': {
+        mergeKey: 'seriesInstanceUid',
+        level: 0,
+        tagFunc: (series, sourceName) => {
+          series.forEach(series => {
+            series.RetrieveAETitle = sourceName;
+            DicomMetadataStore.updateSeriesMetadata(series);
+          });
+          return series;
+        },
+      },
+    };
+
+    const { mergeKey, level, tagFunc } = mergeMap[path] || { tagFunc: x => x, level: -1 };
+
+    const dataSourceDefs = Object.values(extensionManager.dataSourceDefs);
+    const promises = [];
+    let globalData = [];
+
+    for (const dataSourceDef of dataSourceDefs) {
+      const { configuration, sourceName } = dataSourceDef;
+      if (configuration && dataSourceNames.includes(sourceName)) {
+        const [dataSource] = extensionManager.getDataSources(sourceName);
+        const promise = executeFunction(dataSource, path, args);
+        promises.push(promise);
+
+        const getDedupedData = ({ globalData, incomingData, mergeKey, level }) => {
+          const keys = globalData.map(r => getKeyByLevel(r, mergeKey, level));
+          return incomingData.filter(r => !keys.includes(getKeyByLevel(r, mergeKey, level)));
+        };
+
+        const processDedupedData = ({ sourceName, globalData, incomingData, mergeKey, level }) => {
+          let data = [];
+          if (mergeKey && level !== -1) {
+            const dedupedData = getDedupedData({
+              globalData: globalData.flat(),
+              incomingData,
+              mergeKey,
+              level,
+            });
+            data = data.concat(dedupedData);
+          } else {
+            data = data.concat(data);
```

Comment:
should this be `data = data.concat(incomingData)`?

Reply:
Updated, thanks!